### DISCUSSION TITLE
msfupdate: Can't find the 'libpq-fe.h header

### DISCUSSION THREAD

**Discussion by Duffy665** [rapid7/metasploit-framework#15775] [Q&A] [answered]
I hit msfupdate, and this is essentially what happens. I have not been able to get Metasploit to function for a while now. 

`msfupdate
[*]
[*] Attempting to update the Metasploit Framework...
[*]

[*] Checking for updates via git
[*] Note: Updating from bleeding edge
HEAD is now at d5cbb26f09 Bump version of framework to 6.1.11
Already on 'master'
Your branch is ahead of 'origin/master' by 43 commits.
  (use "git push" to publish your local commits)
Already up to date.
[*] Updating gems...
[DEPRECATED] `Bundler.with_clean_env` has been deprecated in favor of `Bundler.with_unbundled_env`. If you instead want the environment before bundler was originally loaded, use `Bundler.with_original_env` (called at /usr/local/bin/msfupdate:231)
Fetching gem metadata from https://rubygems.org/.........
Using rake 13.0.6
Using Ascii85 1.1.0
Using concurrent-ruby 1.0.5
Using i18n 1.8.10
Using minitest 5.14.4
Using tzinfo 2.0.4
Using zeitwerk 2.4.2
Using activesupport 6.1.4.1
Using builder 3.2.4
Using erubi 1.10.0
Using mini_portile2 2.6.1
Using racc 1.5.2
Using nokogiri 1.12.5 (x86_64-darwin)
Using rails-dom-testing 2.0.3
Using crass 1.0.6
Using loofah 2.12.0
Using rails-html-sanitizer 1.4.2
Using actionview 6.1.4.1
Using rack 2.2.3
Using rack-test 1.1.0
Using actionpack 6.1.4.1
Using activemodel 6.1.4.1
Using activerecord 6.1.4.1
Using public_suffix 4.0.6
Using addressable 2.8.0
Using afm 0.2.2
Using arel-helpers 2.12.1
Using ast 2.4.2
Using aws-eventstream 1.2.0
Using aws-partitions 1.515.0
Using aws-sigv4 1.4.0
Using jmespath 1.4.0
Using aws-sdk-core 3.121.1
Using aws-sdk-ec2 1.270.0
Using aws-sdk-iam 1.61.0
Using aws-sdk-kms 1.49.0
Using aws-sdk-s3 1.103.0
Using bcrypt 3.1.16
Using bcrypt_pbkdf 1.1.0
Using bindata 2.4.10
Using bson 4.12.1
Using bundler 2.1.4
Using byebug 11.1.3
Using coderay 1.1.3
Using cookiejar 0.3.3
Using daemons 1.4.1
Using diff-lcs 1.4.4
Using unf_ext 0.0.8
Using unf 0.1.4
Using simpleidn 0.2.1
Using dnsruby 1.61.7
Using docile 1.4.0
Using domain_name 0.5.20190701
Using ed25519 1.2.4
Using eventmachine 1.2.7
Using em-socksify 0.3.2
Using http_parser.rb 0.8.0
Using em-http-request 1.1.7
Using factory_bot 6.2.0
Using method_source 1.0.0
Using thor 1.1.0
Using railties 6.1.4.1
Using factory_bot_rails 6.2.0
Using faker 2.19.0
Using faraday-em_http 1.0.0
Using faraday-em_synchrony 1.0.0
Using faraday-excon 1.1.0
Using faraday-httpclient 1.0.1
Using faraday-net_http 1.0.1
Using faraday-net_http_persistent 1.2.0
Using faraday-patron 1.0.0
Using faraday-rack 1.0.0
Using multipart-post 2.1.1
Using ruby2_keywords 0.0.5
Using faraday 1.8.0
Using websocket-extensions 0.1.5
Using websocket-driver 0.7.5
Using faye-websocket 0.11.1
Using ffi 1.15.4
Using filesize 0.2.0
Using fivemat 1.3.7
Using gssapi 1.3.1
Using gyoku 1.3.1
Using hashery 2.1.2
Using hrr_rb_ssh 0.4.2
Using hrr_rb_ssh-ed25519 0.4.2
Using http-cookie 1.0.4
Using httpclient 2.8.3
Using io-console 0.5.9
Using reline 0.2.5
Using irb 1.3.6
Using rkelly-remix 0.0.7
Using jsobfu 0.4.2
Using json 2.6.0
Using little-plugger 1.1.4
Using multi_json 1.15.0
Using logging 2.3.0
Using memory_profiler 1.0.0
Using metasm 1.0.5
Using metasploit-concern 4.0.3
Using metasploit-model 4.0.3
Fetching pg 1.2.3
Installing pg 1.2.3 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /Users/codydufner/.rvm/gems/ruby-2.7.2/gems/pg-1.2.3/ext
/Users/codydufner/.rvm/rubies/ruby-2.7.2/bin/ruby -I /Users/codydufner/.rvm/rubies/ruby-2.7.2/lib/ruby/2.7.0 -r ./siteconf20211017-32265-kx3xjs.rb extconf.rb
checking for pg_config... no
No pg_config... trying anyway. If building fails, please try again with
 --with-pg-config=/path/to/pg_config
checking for libpq-fe.h... no
Can't find the 'libpq-fe.h header
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/Users/codydufner/.rvm/rubies/ruby-2.7.2/bin/$(RUBY_BASE_NAME)
	--with-pg
	--without-pg
	--enable-windows-cross
	--disable-windows-cross
	--with-pg-config
	--without-pg-config
	--with-pg_config
	--without-pg_config
	--with-pg-dir
	--without-pg-dir
	--with-pg-include
	--without-pg-include=${pg-dir}/include
	--with-pg-lib
	--without-pg-lib=${pg-dir}/lib

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /Users/codydufner/.rvm/gems/ruby-2.7.2/extensions/x86_64-darwin-20/2.7.0/pg-1.2.3/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /Users/codydufner/.rvm/gems/ruby-2.7.2/gems/pg-1.2.3 for inspection.
Results logged to /Users/codydufner/.rvm/gems/ruby-2.7.2/extensions/x86_64-darwin-20/2.7.0/pg-1.2.3/gem_make.out

An error occurred while installing pg (1.2.3), and Bundler cannot continue.
Make sure that `gem install pg -v '1.2.3' --source 'https://rubygems.org/'` succeeds before bundling.

In Gemfile:
  metasploit-framework was resolved to 6.1.11, which depends on
    metasploit-credential was resolved to 5.0.5, which depends on
      metasploit_data_models was resolved to 5.0.4, which depends on
        pg`

**Comment by bcoles** [comment, marked as answer]
> checking for libpq-fe.h... no
> Can't find the 'libpq-fe.h header

Postgresql headers are missing.

For Ubuntu, install `libpq-dev` with `apt install libpq-dev`.

For OSX, perhaps installing postgres will help. `brew install postgresql`.

* https://stackoverflow.com/a/6040822
* https://blog.8bitzen.com/posts/02-10-2020-ubuntu-20.04-resolving--cant-find-the--libpq-fe.h-header-when-trying-to-install-pg-gem

**Reply by Duffy665** [reply]
Thanks ! That fixed it. You are awsome
